### DISCUSSION TITLE
difference between "using" and "#define"

### DISCUSSION THREAD

**Discussion by hpssjellis** [hsaturn/TinyMqtt#10] [Q&A] [answered]
@hsaturn  Can you explain the difference between these lines of code:  [reference lines of code here](https://github.com/hsaturn/TinyMqtt/blob/90dea36ab07090c7b6aa137c1a9c80e1628127ba/src/TinyMqtt.h#L40-L42)

```
  using TcpClient = WiFiClient;
  using TcpServer = WiFiServer;
  //#define TcpClient  WiFiClient     // also compiles but gets a similar error
  //#define TcpServer  WiFiServer

```

I have looked it up but I am used to code like  ```using namespace mbed;```  

For the Portenta neither one works, both give a hard fault. Wondering if it is an issue with how the MBED WiFiClient is written. 

The error is:
```
Exception in thread "Thread-74" java.util.ConcurrentModificationException
	at java.util.LinkedList$ListItr.checkForComodification(LinkedList.java:966)
	at java.util.LinkedList$ListItr.next(LinkedList.java:888)


````

Which typically would be for messing with a linked list inside a "for" loop or the equivalent.

**Comment by hsaturn** [comment, marked as answer]
Hi Jeremy (Lost my answer GRRRR)

#define is just a preprocessing string replacement (not seen by the compiler)

using is a type alias interpreted by the compiler (not the preprocessor)

An example where using cannot be used : [https://onlinegdb.com/MzKsCVN5p](https://onlinegdb.com/MzKsCVN5p)

An example where #define produces bad things [https://www.onlinegdb.com/Mo5xRNL0G](https://www.onlinegdb.com/Mo5xRNL0G)

#define replaces the string in ANY context (except inside strings), while USING uses the alias only within type context.

A last example showing why using is better when used for types [https://onlinegdb.com/drGp8aWKc](https://onlinegdb.com/drGp8aWKc
)

using TcpClient = WiFiClient;    // ==> Tells the compiler to use WifiClient as real class when TcpClient is encountered.

```using namespace mbed;``` Is another usage of using. It says there that we are using the namespace mbed.
This allows to remove mbed:: in front of anything that belongs to mbed.
But this can also leads to conflicts.

```
#include <iostream>
#using namespace std;

void foo() { std::cout << "Hello" << std::endl; }       // <= will compile with and without the #using name space std line

void bar() { cout << "Hello" << endl; }    // will compile only with the using namespace line
```

**Reply by hpssjellis** [reply]
Wow the [online GDB compiler](https://www.onlinegdb.com/) is a great tool. Thanks for such a good explanation.

**Comment by hsaturn** [comment]
```
Exception in thread "Thread-74" java.util.ConcurrentModificationException
	at java.util.LinkedList$ListItr.checkForComodification(LinkedList.java:966)
	at java.util.LinkedList$ListItr.next(LinkedList.java:888)
```

Hi Jeremy

I really cannot figure out why you are giving me many "JAVA" Exceptions stacks.

TinyMqtt is a C++ library, nothing related to Java.

I guess that these error are related to your IDE.

**Reply by hpssjellis** [reply]
LOL, I was only looking at the error, but woke up in the middle of the night asking that same question, then when back to sleep as I didn't know the answer. Perhaps the code works fine just dies in whatever part of the Arduino compiler hits Java.  I might try [platformio.org](https://platformio.org/) what compiler do you use? Arduino regular IDE works fine for my Node MCU 32S.
